python_modules/libraries/dagster-airbyte/dagster_airbyte/managed/reconciliation.py

```diff
@@ -0,0 +1,510 @@
+    diff = diff_dicts(
+        config_dst.destination_configuration if config_dst else {},
+        curr_dst.destination_configuration if curr_dst else {},
+    )
+    if not diff.is_empty():
+        name = config_dst.name if config_dst else curr_dst.name if curr_dst else "Unknown"
+        return ManagedElementDiff().with_nested(name, diff)
+
+    return ManagedElementDiff()
+
+
+def conn_dict(conn: Optional[AirbyteConnection]) -> Dict[str, Any]:
+    if not conn:
+        return {}
+    return {
+        "source": conn.source.name if conn.source else "Unknown",
+        "destination": conn.destination.name if conn.destination else "Unknown",
+        "normalize data": conn.normalize_data,
+        "streams": {k: v.name for k, v in conn.stream_config.items()},
+    }
+
+
+def diff_connections(
+    config_conn: Optional[AirbyteConnection], curr_conn: Optional[AirbyteConnection]
+) -> ManagedElementCheckResult:
+    """
+    Utility to diff two AirbyteConnection objects.
+    """
+    diff = diff_dicts(conn_dict(config_conn), conn_dict(curr_conn))
+    if not diff.is_empty():
+        name = config_conn.name if config_conn else curr_conn.name if curr_conn else "Unknown"
+        return ManagedElementDiff().with_nested(name, diff)
+
+    return ManagedElementDiff()
+
+
+def reconcile_sources(
+    res: AirbyteResource,
+    config_sources: Mapping[str, AirbyteSource],
+    existing_sources: Mapping[str, InitializedAirbyteSource],
+    workspace_id: str,
+    dry_run: bool,
+) -> Tuple[Mapping[str, InitializedAirbyteSource], ManagedElementCheckResult]:
+    """
+    Generates a diff of the configured and existing sources and reconciles them to match the
+    configured state if dry_run is False.
+    """
+
+    diff = ManagedElementDiff()
+
+    initialized_sources = {}
+    for source_name in set(config_sources.keys()).union(existing_sources.keys()):
+        configured_source = config_sources.get(source_name)
+        existing_source = existing_sources.get(source_name)
+
+        diff = diff.join(
+            diff_sources(configured_source, existing_source.source if existing_source else None)
+        )
+
+        if existing_source and (
+            not configured_source or (configured_source.must_be_recreated(existing_source.source))
+        ):
+            initialized_sources[source_name] = existing_source
+            if not dry_run:
+                res.make_request(
+                    endpoint="/sources/delete",
+                    data={"sourceId": existing_source.source_id},
+                )
+            existing_source = None
+
+        if configured_source:
+            defn_id = check.not_none(
+                res.get_source_definition_by_name(configured_source.source_type, workspace_id)
+            )
+            base_source_defn_dict = {
+                "name": configured_source.name,
+                "connectionConfiguration": configured_source.source_configuration,
+            }
+            source_id = "TBD"
```

Comment:
nit: maybe make this a bit more explicit, like "<NOT YET CREATED>", just to visually distinguish it (also applies to the destination_id in the other function)